File: src/mon_agent_server/tools/memory.py
from __future__ import annotations

import asyncio
import re
from typing import Any

from mon_agent_core import AgentTool

from .context import MonToolContext
from .core_access import core_call, require_core_access
from .datetime_utils import format_local_datetime
from .result import text_result, tool_failure


_SECRET_PATTERN = re.compile(
    r"(?:sk-[A-Za-z0-9_-]{16,}|(?:api[_ -]?key|token|password|密码|密钥)\s*[:=：]\s*\S+)",
    re.IGNORECASE,
)


def _require_root_writer(context: MonToolContext) -> None:
    if context.agent_path != "/root":
        raise tool_failure("blocked", "子智能体只能检索长期记忆；写入、修改和遗忘必须由父智能体执行。")


def _safe_content(value: Any) -> str:
    content = re.sub(r"\s+", " ", str(value or "").strip())
    if not content:
        raise ValueError("记忆内容不能为空。")
    if _SECRET_PATTERN.search(content):
        raise ValueError("检测到可能的密钥、令牌或密码，拒绝写入长期记忆。")
    return content


def _memory_line(memory: dict[str, Any]) -> str:
    created_raw = memory.get("created_at")
    updated_raw = memory.get("updated_at")
    created_text = format_local_datetime(created_raw)
    updated_text = format_local_datetime(updated_raw)
    lines = [
        f"#{memory.get('id')} [{memory.get('kind') or 'fact'}]",
        f"   写入时间（本地）: {created_text}",
    ]
    if updated_raw and updated_text != created_text:
        lines.append(f"   更新时间（本地）: {updated_text}")
    lines.append(f"   内容: {memory.get('content') or ''}")
    return "\n".join(lines)


def _agent_character_id(context: MonToolContext, params: dict[str, Any]) -> int | str | None:
    explicit = params.get("agent_character_id")
    if explicit not in (None, ""):
        return explicit
    character = context.character if isinstance(context.character, dict) else {}
    return character.get("id")


def create_memory_tools(context: MonToolContext) -> list[AgentTool]:
    async def remember_execute(_call_id: str, params: dict[str, Any], _signal: Any = None, _update: Any = None):
        _require_root_writer(context)
        core, token = require_core_access(context)
        scope_type = params.get("scope_type") or "agent_character"
        agent_character_id = _agent_character_id(context, params)
        if scope_type == "agent_character" and not agent_character_id:
            raise ValueError("智能体角色记忆必须提供 agent_character_id，或在当前会话绑定角色。")
        memory = await asyncio.to_thread(
            core_call,
            core.remember_memory,
            token,
            {
                "content": _safe_content(params["content"]),
                "kind": params.get("kind") or "fact",
                "scope_type": scope_type,
                "scope_key": str(agent_character_id or "") if scope_type == "agent_character" else params.get("scope_key") or "",
                "assistant": params.get("assistant_id"),
                "agent_character": agent_character_id if scope_type == "agent_character" else None,
                "source_session_id": context.session_id or "",
                "source_message_ids": [context.get_message_id()] if context.get_message_id and context.get_message_id() else [],
                "metadata": {"source": "explicit_tool", "agent_path": context.agent_path},
            },
        )
        return text_result(f"已写入长期记忆：\n\n{_memory_line(memory)}", {"memory": memory})

    async def search_execute(_call_id: str, params: dict[str, Any], _signal: Any = None, _update: Any = None):
        core, token = require_core_access(context)
        memories = await asyncio.to_thread(
            core_call,
            core.list_memories,
            token,
            {
                "q": params.get("query"),
                "scope_type": params.get("scope_type"),
                "scope_key": params.get("scope_key"),
                "kind": params.get("kind"),
                "assistant": params.get("assistant_id"),
                "agent_character": _agent_character_id(context, params),
                "limit": params.get("limit") or 10,
            },
        )
        body = "没有找到相关长期记忆。" if not memories else "相关长期记忆：\n\n" + "\n".join(
            f"- {_memory_line(memory)}" for memory in memories
        )
        return text_result(body, {"memories": memories, "count": len(memories)})

    async def update_execute(_call_id: str, params: dict[str, Any], _signal: Any = None, _update: Any = None):
        _require_root_writer(context)
        core, token = require_core_access(context)
        memory = await asyncio.to_thread(
            core_call,
            core.update_memory,
            token,
            int(params["id"]),
            {"content": _safe_content(params["content"]), **({"kind": params["kind"]} if params.get("kind") else {})},
        )
        return text_result(f"已更新长期记忆：\n\n{_memory_line(memory)}", {"memory": memory})

    async def forget_execute(_call_id: str, params: dict[str, Any], _signal: Any = None, _update: Any = None):
        _require_root_writer(context)
        core, token = require_core_access(context)
        memory = await asyncio.to_thread(core_call, core.forget_memory, token, int(params["id"]))
        return text_result(f"已遗忘长期记忆 #{memory.get('id')}。", {"memory": memory})

    scope_properties = {
        "scope_type": {"type": "string", "enum": ["agent_character", "project", "workspace"]},
        "scope_key": {"type": "string"},
        "assistant_id": {"type": "number"},
        "agent_character_id": {"type": "number"},
    }
    return [
        AgentTool(
            "remember_memory", "记住长期信息",
            "当用户明确要求记住，或确认了未来会持续有用的稳定偏好、事实、决策或流程时写入当前智能体角色的长期记忆。角色之间不共享记忆。不要保存密码、密钥、临时进度或猜测。",
            {"type": "object", "properties": {"content": {"type": "string"}, "kind": {"type": "string", "enum": ["preference", "fact", "decision", "procedure"]}, **scope_properties}, "required": ["content"]},
            remember_execute, execution_mode="sequential",
        ),
        AgentTool(
            "search_memories", "搜索长期记忆",
            "按需搜索当前用户可见的历史偏好、事实、决策和流程；普通相关记忆会由运行时自动召回。",
            {"type": "object", "properties": {"query": {"type": "string"}, "kind": {"type": "string"}, "limit": {"type": "number"}, **scope_properties}},
            search_execute,
        ),
        AgentTool(
            "update_memory", "更新长期记忆",
            "修正一条已知 ID 的长期记忆。",
            {"type": "object", "properties": {"id": {"type": "number"}, "content": {"type": "string"}, "kind": {"type": "string"}}, "required": ["id", "content"]},
            update_execute, execution_mode="sequential",
        ),
        AgentTool(
            "forget_memory", "遗忘长期记忆",
            "在用户明确要求遗忘且目标记忆 ID 已确定时使用。目标不明确时先搜索或询问。",
            {"type": "object", "properties": {"id": {"type": "number"}}, "required": ["id"]},
            forget_execute, execution_mode="sequential",
        ),
    ]

File: crates/eden-agent-host/src/tools/memory.rs
use crate::{
    HostServices,
    support::{integer, limit, output, store_failure, string},
};
use async_trait::async_trait;
use eden_agent_core::{
    PermissionRequest, Tool, ToolCall, ToolCallContext, ToolDefinition, ToolFailure, ToolOutput,
};
use eden_agent_domain::SessionId;
use eden_agent_store::Store;
use regex::Regex;
use serde_json::{Value, json};
use std::sync::Arc;

#[derive(Clone, Copy)]
enum MemoryAction {
    Remember,
    Search,
    Update,
    Forget,
}
impl MemoryAction {
    const ALL: [Self; 4] = [Self::Remember, Self::Search, Self::Update, Self::Forget];
}
struct MemoryTool {
    host: HostServices,
    action: MemoryAction,
}

#[async_trait]
impl Tool for MemoryTool {
    fn definition(&self) -> ToolDefinition {
        let (name, description, required) = match self.action {
            MemoryAction::Remember => (
                "remember_memory",
                "Store durable long-term memory",
                vec!["content"],
            ),
            MemoryAction::Search => ("search_memories", "Search long-term memory", vec![]),
            MemoryAction::Update => (
                "update_memory",
                "Correct long-term memory",
                vec!["id", "content"],
            ),
            MemoryAction::Forget => ("forget_memory", "Delete long-term memory", vec!["id"]),
        };
        let mut definition = ToolDefinition::direct(name, description);
        definition.parameters = json!({"type":"object","required":required,"properties":{"id":{"type":"integer"},"content":{"type":"string"},"query":{"type":"string"},"kind":{"type":"string","enum":["preference","fact","decision","procedure"]},"limit":{"type":"integer"}}});
        definition
    }
    fn permission_request(&self, arguments: &Value) -> Option<PermissionRequest> {
        if matches!(self.action, MemoryAction::Search) {
            None
        } else {
            Some(PermissionRequest {
                permission: "memory.write".to_owned(),
                patterns: vec![
                    arguments
                        .get("id")
                        .map_or_else(|| "new".to_owned(), Value::to_string),
                ],
                always: vec!["*".to_owned()],
            })
        }
    }
    async fn execute(
        &self,
        call: &ToolCall,
        context: ToolCallContext,
    ) -> Result<ToolOutput, ToolFailure> {
        let args = &call.arguments;
        if !matches!(self.action, MemoryAction::Search) && is_subagent(&context) {
            return Err(ToolFailure::new(
                "blocked",
                "子智能体只能检索长期记忆；写入、修改和遗忘必须由主智能体执行",
            ));
        }
        let (scope_type, scope_key) = memory_scope(&self.host.store, &context).await?;
        let result = match self.action {
            MemoryAction::Remember => {
                let content = safe_memory(&string(args, "content")?)?;
                serde_json::to_value(
                    self.host
                        .store
                        .create_memory(
                            &content,
                            args.get("kind").and_then(Value::as_str).unwrap_or("fact"),
                            &scope_type,
                            &scope_key,
                            context.session_id.as_deref().unwrap_or(""),
                            json!({"source":"explicit_tool","agentCharacterId":scope_key}),
                        )
                        .await
                        .map_err(store_failure)?,
                )
                .unwrap_or_default()
            }
            MemoryAction::Search => serde_json::to_value(
                self.host
                    .store
                    .search_memories_in_scope(
                        &scope_type,
                        &scope_key,
                        args.get("query").and_then(Value::as_str),
                        limit(args),
                    )
                    .await
                    .map_err(store_failure)?,
            )
            .unwrap_or_default(),
            MemoryAction::Update => {
                let content = safe_memory(&string(args, "content")?)?;
                ensure_memory_scope(
                    &self.host.store,
                    integer(args, "id")?,
                    &scope_type,
                    &scope_key,
                )
                .await?;
                serde_json::to_value(
                    self.host
                        .store
                        .update_memory(
                            integer(args, "id")?,
                            &content,
                            args.get("kind").and_then(Value::as_str),
                        )
                        .await
                        .map_err(store_failure)?,
                )
                .unwrap_or_default()
            }
            MemoryAction::Forget => {
                ensure_memory_scope(
                    &self.host.store,
                    integer(args, "id")?,
                    &scope_type,
                    &scope_key,
                )
                .await?;
                self.host
                    .store
                    .delete_memory(integer(args, "id")?)
                    .await
                    .map_err(store_failure)?;
                json!({"deleted":true})
            }
        };
        Ok(output(result))
    }
}

fn is_subagent(context: &ToolCallContext) -> bool {
    context
        .metadata
        .get("agentPath")
        .and_then(Value::as_str)
        .is_some_and(|path| !path.trim().is_empty() && path != "/root")
}

async fn memory_scope(
    store: &Store,
    context: &ToolCallContext,
) -> Result<(String, String), ToolFailure> {
    let session_id = context
        .session_id
        .as_deref()
        .ok_or_else(|| ToolFailure::new("session_required", "长期记忆操作需要当前会话"))?
        .parse::<SessionId>()
        .map_err(|error| ToolFailure::new("invalid_session", error.to_string()))?;
    let session = store.get_session(session_id).await.map_err(store_failure)?;
    let character_id = session
        .participants
        .first()
        .and_then(|participant| scalar_id(participant.get("characterId")))
        .or_else(|| {
            session
                .participants
                .first()
                .and_then(|participant| participant.get("profile"))
                .and_then(|profile| profile.get("character"))
                .and_then(|character| scalar_id(character.get("id")))
        })
        .ok_or_else(|| {
            ToolFailure::new(
                "character_required",
                "当前会话没有绑定角色，不能访问角色长期记忆",
            )
        })?;
    Ok(("agent_character".to_owned(), character_id))
}

async fn ensure_memory_scope(
    store: &Store,
    id: i64,
    scope_type: &str,
    scope_key: &str,
) -> Result<(), ToolFailure> {
    let memory = store.get_memory(id).await.map_err(store_failure)?;
    if memory.scope_type != scope_type || memory.scope_key != scope_key {
        return Err(ToolFailure::new(
            "memory_scope_denied",
            "目标记忆不属于当前角色",
        ));
    }
    Ok(())
}

fn scalar_id(value: Option<&Value>) -> Option<String> {
    match value? {
        Value::String(value) if !value.trim().is_empty() => Some(value.trim().to_owned()),
        Value::Number(value) => Some(value.to_string()),
        _ => None,
    }
}

fn safe_memory(content: &str) -> Result<String, ToolFailure> {
    let content = content.split_whitespace().collect::<Vec<_>>().join(" ");
    let secrets =
        Regex::new(r"(?i)(sk-[A-Za-z0-9_-]{16,}|(?:api[_ -]?key|token|password|密码|密钥|令牌)\s*[:=：]\s*\S+)")
            .expect("valid regex");
    if secrets.is_match(&content) {
        Err(ToolFailure::new(
            "secret_rejected",
            "credentials cannot be stored in long-term memory",
        ))
    } else {
        Ok(content)
    }
}

pub(super) fn tools(host: HostServices) -> Vec<Arc<dyn Tool>> {
    MemoryAction::ALL
        .iter()
        .copied()
        .map(|action| {
            Arc::new(MemoryTool {
                host: host.clone(),
                action,
            }) as Arc<dyn Tool>
        })
        .collect()
}
